### PR TITLE
Add Trello Module with API and Test

### PR DESCRIPTION
**Pull request by Sidhved**
# Add Trello Module

This PR adds a new Trello module to the project, providing a clean API interface for interacting with Trello boards and cards.

### New Features
- Add TrelloManager class for direct Trello API interactions
- Add public API interface for common Trello operations
- Implement comprehensive test suite with proper mocking
- Add type hints throughout the module

### Specific Implementations
- Card creation and management
- Checklist operations
- List management
- Board operations
- Comment and label handling

### Code Quality
- Add mypy configuration for type checking
- Fix linting issues
- Improve code documentation
- Add proper error handling
- Implement singleton pattern for TrelloManager

### Testing
- Add unit tests for all API functions
- Add tests for TrelloManager
- Implement proper mocking of Trello API calls
- Add test fixtures for common test scenarios

## Testing Done
- All unit tests pass
- mypy type checking passes
- Ruff linting passes
- Manually tested core functionality
image

## Team Members
@Sidhved and @cindyke124

### CODE REVIEW

File: src/maigic_nyu/trello/_trello_manager.py [hunk 1]
```diff
@@ -0,0 +1,405 @@
...
+class TrelloManager:
+    """A class to interact with the Trello API."""
+
+    def __init__(self) -> None:
+        """Initialize TrelloManager with API credentials."""
+        load_dotenv()
+        self.api_key = os.getenv("TRELLO_API_KEY")
+        self.token = os.getenv("TRELLO_OAUTH_TOKEN")
+        self.BASE_URL = "https://trello.com/invite/b/673e7c539b668d5238d14115/ATTI9797fcaf204991af7e18fa9e0c3a421f24EC6638/maigictest"
+
```
Comment (isiddharthsingh) [inline src/maigic_nyu/trello/_trello_manager.py:65]: The Base_URL is exposed to public. Suggestion to add it in .env file.

**Comment by AM-git-hub** [comment]
### 1. Change Granularity
**Observation**: The pull request focuses solely on implementing a Trello module with API functionality and its corresponding tests. The changes are well-contained within the scope of Trello API interaction. However, more incremental commits with descriptive messages would improve tracking of changes.

### 2. Code Functionality
**Does the Code Meet Requirements?**: Yes, the module achieves its goal of enabling interaction with the Trello API for board and card management.
However, we could not find the api caller (slack bot?) that would make these APIs usable. (Are you using the terminal to interact with the Trello bot?)

**Logic Flow**: The logic for API calls and data handling is clear, but error handling could be improved. For instance:
- API responses are not consistently checked for success (e.g., using response.status_code).
- Exceptions for failed API calls should be explicitly managed and reported.


### 3. Readability & Structure
**Naming Conventions**: Most variable and function names are descriptive, but certain abbreviations (e.g., brd_id) could be replaced with clearer names like board_id.
**Modularity**: The separation of Trello API interactions into a dedicated module is commendable.
However, utility functions such as create_board and add_card_to_board could be refactored into a helper class or reused across methods to reduce redundancy.

### 4. Standards Compatibility
**Linting**: Code appears mostly aligned with PEP 8 standards, but minor issues (e.g., line length exceeding 79 characters) could be resolved with a linter like flake8 or black.
**API Exposure**: Hardcoded API keys and tokens in the code present a major security concern. It is essential to migrate these to environment variables using libraries like python-dotenv. This approach will also support secure deployment practices.

### 5. Testing
**Coverage**: Unit tests are included, covering basic functionality like creating boards and adding cards. However, there is a lack of tests for edge cases, such as handling invalid API keys or network failures.
**Readability**: The test cases are easy to read and follow. Adding comments explaining the purpose of each test would further enhance clarity.
Mocking is used correctly to avoid actual API calls, which is a best practice for external API testing.

**Reviewed by:**
1) Adittya Mittal - am14079
2) Siddharth Singh - sms10221

File: src/maigic_nyu/trello/api.py [hunk 2]
```diff
@@ -0,0 +1,247 @@
+"""API module that exposes Trello functionality to users."""
```
Comment (kyotov) [inline src/maigic_nyu/trello/api.py:1]: Why is this outside `trello` package?

Comment (kyotov) [reply src/maigic_nyu/trello/api.py:1]: seems duplicated!?

File: src/maigic_nyu/trello/api.py [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+
+# Global TrelloManager instance
+_trello = None
+
+def _get_trello() -> TrelloManager:
+    """Get or create the global TrelloManager instance."""
+    global _trello
+    if _trello is None:
+        _trello = TrelloManager()
+    return _trello
```
Comment (kyotov) [inline src/maigic_nyu/trello/api.py:48]: this singleton pattern is better implemented in a class where the global variable is a `ClassVar`

**Review by kyotov — CHANGES_REQUESTED** [review]
some minor comments -- but this is getting there.

the general list i am sending to everybody is 

* How is each file in your PR connected to the task at hand?  
  * You should not have files that are irrelevant to the task.  
* Make sure files that should be .gitignore are in .gitignore  
  * E.g. .DS\_Store  
* Make sure secrets are not checked in\!  
* Make sure there is a clear and separated API\!

You have some of these right, but not all.
Make sure you follow all of them.

Then I will look again.
